packages/common/src/hooks/useImageSize2.ts

```diff
@@ -0,0 +1,134 @@
+import { useState, useEffect, useCallback } from 'react'
+
+import { SquareSizes, Track } from '~/models'
+import { Maybe } from '~/utils/typeUtils'
+
+// Global image cache
+const IMAGE_CACHE = new Set<string>()
+
+/**
+ * Fetches an image from the given artwork object managing sizes and using fallback mirrors if necessary.
+ *  - If a larger image has already been fetched and is in the case, use it instead
+ *  - If a smaller image has already been fetched and is in the cache, use it while fetching the larger one
+ *  - If fetching a given image fails, substitute out mirrors and try again
+ * @param artwork - The artwork object to fetch from
+ * @param targetSize - The desired size of the image
+ * @param defaultImage - The fallback image to use if no image is found in the `artwork` object
+ * @returns The url of the image, or undefined if the image is not available
+ */
+export const useImageSize2 = ({
+  artwork,
+  targetSize,
+  defaultImage
+}: {
+  artwork: Maybe<Track['artwork']>
+  targetSize: SquareSizes
+  defaultImage: string
+}) => {
+  const [imageUrl, setImageUrl] = useState<string | null>(null)
+
+  const preloadImage = (url: string) => {
+    return new Promise((resolve, reject) => {
+      const img = new Image()
+      img.src = url
+      img.onload = () => resolve(url)
+      img.onerror = () => reject(url)
+    })
+  }
+
+  const fetchWithFallback = useCallback(
+    async (url: string) => {
+      const mirrors = [...(artwork?.mirrors ?? [])]
+      let currentUrl = url
+
+      while (mirrors.length > 0) {
+        try {
+          await preloadImage(currentUrl)
+          return currentUrl
+        } catch {
+          const nextMirror = mirrors.shift()
+          if (!nextMirror) throw new Error('No mirror found')
+          const currentHost = new URL(currentUrl).hostname
+          const nextHost = new URL(nextMirror).hostname
+          const newUrl = currentUrl.replace(currentHost, nextHost)
+          currentUrl = newUrl
```

Comment:
A couple of considerations for logic like this:
* You can write directly to `hostname` of a URL instead of using `replace` (`newUrl.hostname = nextHost`). But since we're starting with a string, maybe replace is fine? I just get nervous about string replace stuff instead of using classes that are specifically designed to manipulate a URL.
* Just checking that we don't ever expect any of these urls to contain port numbers? (thinking specifically of local dev). If we _do_expect that, it might be worth considering usage of `host` so that you get hostname+any specified port. And `host` is also a writable field
* Given that this logic is slightly tricky, might be worth extracting it out to a helper that is shared with the streamUrl code so that we are using the same process everywhere for host substitution?

Reply:
for now simplified to
```
const nextUrl = new URL(currentUrl)
nextUrl.hostname = new URL(nextMirror).hostname
currentUrl = nextUrl.toString()
```

I think it's excessive to share this logic, and I think better to have this image fetch code independent from that code entirely. i think that's one of the reasons our current image stuff got so complex. very overloaded